test-data/unit/check-typeddict.test

```diff
@@ -2590,3 +2590,30 @@ TD[str](key=0, value=0)  # E: Incompatible types (expression has type "int", Typ
 TD[str]({"key": 0, "value": 0})  # E: Incompatible types (expression has type "int", TypedDict item "value" has type "str")
 [builtins fixtures/dict.pyi]
 [typing fixtures/typing-typeddict.pyi]
+
+[case testTypedDictSelfTypeNewSyntaxGeneric]
+# flags: --strict-optional
+from typing import Self, TypedDict, Optional, Generic, TypeVar
+
+T = TypeVar("T")
+class TD(TypedDict, Generic[T]):
+    val: T
+    next: Optional[Self]
```

Comment:
What happens if we inherit a typed dict from `TD`?

Reply:
TypedDicts can't be inherited in a "normal" sense of this word, since we can't change item types, we can only add new items, so I usually call it extending.

Currently I eagerly expand `Self` in TypedDicts (and in NamedTuples), so in this case `next` will be still `Optional[TD]` in extended TypedDict. PEP doesn't say anything about TypedDicts and NamedTuples, and they are not really classes, so I just implemented whatever looked more natural (and also simpler to implement).

But actually thinking a bit more about this, maybe it would be better to implement them more like for classes, so it will point to extended dict. I am not sure yet.